### PR TITLE
HWKAPM-478 Support caching information in a clustered environment

### PR DESCRIPTION
**Pull request by objectiser**

### CODE REVIEW

**Comment by objectiser** [comment]
Review nodes - this PR contains some minor changes relates to serialization of some events to deal with null strings. The main change is how cached data (Producer Information and Communication Details) is stored. The initial implementation assumed when clustering the servers a distributed cache would be used - however due to problems getting the distributed cache working lead me to rethink this approach - as the data being cached is distributed via JMS anyway, rather than redistribute it again via infinispan, it seemed better to leverage local caches (to help with memory management) that are initialised via the JMS published data.

There are some potential issues, discussed in https://issues.jboss.org/browse/HWKAPM-479, which may mean a distributed cache will eventually be used, but while the main focus is on monitoring short lived transaction/trace instances, this seems like a viable solution. So the main changes are to introduce some non-durable subscribing topic MDBs to receive the data and store it in the local caches.

File: api/src/main/java/org/hawkular/apm/api/utils/SerializationUtil.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+package org.hawkular.apm.api.utils;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+
+/**
+ * @author gbrown
+ */
+public class SerializationUtil {
```
Comment (pilhuhn) [inline api/src/main/java/org/hawkular/apm/api/utils/SerializationUtil.java]: Have you looked into using some other serialization library instead? I recall that the default from the jdk is rather slow and clumsy. 
I don't mean that you need to change it now

File: processors/communicationdetails-deriver/src/test/java/org/hawkular/apm/processor/communicationdetails/ProducerInfoTest.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+public class ProducerInfoTest {
+
+    @Test
+    public void testSerialize() {
+        Set<Property> props = new HashSet<Property>();
+        Property prop1 = new Property();
+        prop1.setName("testProp1");
+        prop1.setValue("testValue1");
+        prop1.setType(PropertyType.Text);
+        props.add(prop1);
```
Comment (pilhuhn) [inline processors/communicationdetails-deriver/src/test/java/org/hawkular/apm/processor/communicationdetails/ProducerInfoTest.java]: I think you should also test your null values from above

File: server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java [hunk 3]
```diff
@@ -30,17 +30,20 @@
...
+ *
  * @author gbrown
  */
-@MessageDriven(name = "CommunicationDetails_Cache", messageListenerInterface = MessageListener.class,
+@MessageDriven(name = "CommunicationDetails_CommunicationDetailsCache", messageListenerInterface = MessageListener.class,
         activationConfig =
         {
                 @ActivationConfigProperty(propertyName = "destinationType", propertyValue = "javax.jms.Topic"),
-                @ActivationConfigProperty(propertyName = "destination", propertyValue = "CommunicationDetails"),
-                @ActivationConfigProperty(propertyName = "subscriptionDurability", propertyValue = "Durable"),
```
Comment (pilhuhn) [inline server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java]: Why are you removing the Durability here?

**Comment by pilhuhn** [comment]
I think it certainly makes sense to pick the data from the Topic, if it is sent there anyway, to update all local caches.

File: api/src/main/java/org/hawkular/apm/api/utils/SerializationUtil.java (hunk 1 above)
Comment (objectiser) [inline api/src/main/java/org/hawkular/apm/api/utils/SerializationUtil.java]: No haven't looked into any others yet, but if can look at others if the performance in this area becomes an issue. Currently the serialization is only used with the caching.

File: processors/communicationdetails-deriver/src/test/java/org/hawkular/apm/processor/communicationdetails/ProducerInfoTest.java (hunk 2 above)
Comment (objectiser) [inline processors/communicationdetails-deriver/src/test/java/org/hawkular/apm/processor/communicationdetails/ProducerInfoTest.java]: Will add.

File: server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java (hunk 3 above)
Comment (objectiser) [inline server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java]: I'll need to check regarding artemis docs - but main reason was to ensure all nodes received the published message. With the current config, the messages are load balanced across subscribers.

Comment (jshaughn) [inline server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java]: Are you saying that not all subscribers (MDBs) to the topic are getting all messages published to the topic?  If that is not true then it breaks what I always believed about topics and we may need to revisit some other code.

**Comment by jshaughn** [comment]
>  The initial implementation assumed when clustering the servers a distributed cache would be used - however due to problems getting the distributed cache working lead me to rethink this approach...

This is also disturbing as we're depending on ispan caches in a couple of other places. Can you elaborate?

File: server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java (hunk 3 above)
Comment (objectiser) [inline server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java]: This is specific behaviour that I required, that is supported by Artemis, where if more than one durable subscribers have the same subscription name and client id, then they only one will receive the message - it is essentially providing load balancing queue behaviour but when subscribing to a topic - defined by the consumer - so topic publisher is oblivious of how the message is being consumed - which also means normal non durable and durable subscribers can receive receive the message on multiple nodes - so the behaviour you would normally expect from a topic subscriber.

**Comment by objectiser** [comment]
I had two wildfly 10 servers running in a cluster, with a distributed cache defined in the standalone-ha.xml. Logs showed the cluster was being established, but for some reason no data was being shared between the nodes.

Its most likely I got some config wrong, but once I thought about the current solution (based on JMS), I didn't investigate further.

File: server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java (hunk 3 above)
Comment (objectiser) [inline server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java]: Have read further in the JMS/wildfly docs - if clientId and subscriptionName are the same on multiple nodes, then they will load balance the messages from the durable subscription (topic). However to get all nodes receiving from the topic, but still be durable, I just need to change the clientId - which can apparently be done in the MDB using "Annotation Property Replacement" :https://docs.jboss.org/author/display/WFLY10/Java+EE+Application+Deployment+Configuration

Will need to experiment to see what the issues might be if (for example) a node name is not specified.

Comment (pilhuhn) [inline server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java]: I think what you write matches what I recall from the past. Having different client ids (e.g. uuid of the WF10 server you are running on) + durable subscription sounds like what we need.

How do you bootstrap the cache?

Comment (objectiser) [inline server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java]: initially it won't be - the data is short lived, so as long as other servers in the cluster have the data, it should be ok (i.e. the request can be handled by a different server)

**Comment by objectiser** [comment]
@pilhuhn Added the extra test, and changed MDBs so now use durable topic subscription based on a clientId associated with the 'jboss.node.name'.

**Comment by objectiser** [comment]
@pilhuhn ok to merge?
